### PR TITLE
Add deck share provider support

### PR DESCRIPTION
**Pull request by juliusknorr**
This PR extends the share manager to allow apps registering their own share providers and implements support for deck shares in places where handing over the logic to the apps is yet not possible. Basically the parts where the server still needs app based logic is mainly the share api and the sharing sidebar in files.

Heavily inspired by https://github.com/nextcloud/server/pull/10255

Server part of https://github.com/nextcloud/deck/pull/2638 This can already be tested together for the share provider logic, however the deck side is still lacking the user interface to interact with the shares and upload/share files from within deck.

### CODE REVIEW

File: lib/private/Share20/ProviderFactory.php [hunk 1]
```diff
@@ -295,6 +318,8 @@ public function getProviderForType($shareType) {
 			$provider = $this->getShareByCircleProvider();
 		} elseif ($shareType === IShare::TYPE_ROOM) {
 			$provider = $this->getRoomShareProvider();
+		} elseif ($shareType === IShare::TYPE_DECK) {
+			$provider = $this->getProvider('deck');
```
Comment (juliushaertl) [inline lib/private/Share20/ProviderFactory.php:322]: We could also extend the share manager to allow registering a share provider together with a shareType, then at least the provider registration could be done without having additional code in server (except for the actual shareType constant), but i kept that for later to not overcomplicate this pr.

Comment (juliushaertl) [reply lib/private/Share20/ProviderFactory.php:322]: 'deck' is the identifier of the share provider https://github.com/nextcloud/deck/pull/2638/files#diff-bfa062609a398b64ebc3167faba3abbca19b04da0c1e3f3dd033fa6babaebbc4R100-R102 that basically implies that `IShare::TYPE_DECK` is used.

File: apps/files/lib/Controller/ApiController.php [hunk 2]
```diff
@@ -221,7 +221,8 @@ private function getShareTypes(Node $node) {
 			IShare::TYPE_LINK,
 			IShare::TYPE_REMOTE,
 			IShare::TYPE_EMAIL,
-			IShare::TYPE_ROOM
+			IShare::TYPE_ROOM,
+			IShare::TYPE_DECK
```
Comment (nickvergessen) [inline apps/files/lib/Controller/ApiController.php:225]:
```suggestion
			IShare::TYPE_DECK,
```

File: apps/files_sharing/lib/Controller/DeletedShareAPIController.php [hunk 3]
```diff
@@ -213,4 +222,21 @@ private function getRoomShareHelper() {
...
+	 *
+	 * @return \OCA\Deck\Sharing\ShareAPIHelper
+	 * @throws QueryException
+	 */
+	private function getDeckShareHelper() {
+		if (!$this->appManager->isEnabledForUser('deck')) {
+			throw new QueryException();
+		}
+
+		return $this->serverContainer->query('\OCA\Deck\Sharing\ShareAPIHelper');
```
Comment (nickvergessen) [inline apps/files_sharing/lib/Controller/DeletedShareAPIController.php:240]:
```suggestion
		return $this->serverContainer->get(\OCA\Deck\Sharing\ShareAPIHelper::class);
```

File: apps/files_sharing/lib/Controller/ShareAPIController.php [hunk 4]
```diff
@@ -1520,6 +1561,22 @@ private function getRoomShareHelper() {
...
+	 *
+	 * @return \OCA\Deck\Sharing\ShareAPIHelper
+	 * @throws QueryException
+	 */
+	private function getDeckShareHelper() {
+		if (!$this->appManager->isEnabledForUser('deck')) {
+			throw new QueryException();
+		}
+
+		return $this->serverContainer->query('\OCA\Deck\Sharing\ShareAPIHelper');
```
Comment (nickvergessen) [inline apps/files_sharing/lib/Controller/ShareAPIController.php:1578]:
```suggestion
		return $this->serverContainer->get(\OCA\Deck\Sharing\ShareAPIHelper::class);
```

File: lib/private/Share/Constants.php [hunk 5]
```diff
@@ -70,6 +70,10 @@ class Constants {
 	 */
 	public const SHARE_TYPE_ROOM = 10;
 	// const SHARE_TYPE_USERROOM = 11; // Internal type used by RoomShareProvider
+	/**
+	 * @deprecated 21.0.0 - use IShare::TYPE_ROOM instead
```
Comment (nickvergessen) [inline lib/private/Share/Constants.php:74]: what?

**Review by nickvergessen — APPROVED** [review]
🐘 looks like it could work. but didnt really test or check if all places have been adapted.

**Review by rullzer — APPROVED** [review]
Please check the comments from @nickvergessen but fine by me

**Pull request merged**